### ISSUE TITLE
False positive `--py3k` `comprehension-escape`

### ISSUE THREAD

**Issue by asottile** [PyCQA/pylint#2130]
### Steps to reproduce

```python
from __future__ import print_function

y = [x for x in range(1)]

for x in range(2):
    print(x)
```

### Current behavior

```
$ pylint --py3k test.py
No config file found, using default configuration
************* Module test
W:  5,10: Using a variable that was bound inside a comprehension (comprehension-escape)

---------------------------------------------------------------------
Your code has been rated at 7.50/10 (previous run: -20.00/10, +27.50)

```

### Expected behavior

No errors

A heuristic (that I think is valid?) is that a name `x` is rebound after the comprehension.

### pylint --version output

```
$ pylint --version
No config file found, using default configuration
pylint 1.9.1, 
astroid 1.6.4
Python 2.7.15 (default, May  1 2018, 16:44:08) 
[GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.1)]
```

**Comment by hvelarde** [comment]
@PCManticore I found another false positive: https://travis-ci.org/plonegovbr/brasil.gov.agenda/builds/391477517#L2004

this is the code: https://github.com/plonegovbr/brasil.gov.agenda/blob/2.0a1/src/brasil/gov/agenda/tests/test_setup.py#L184-L219

and a snippet:

```python
step = [
    step for step in upgradeSteps
    if (step[0]['dest'] == ('2000',)) and (step[0]['source'] == ('1000',))]
```

**Comment by PCManticore** [comment]
Hi @hvelarde With what pylint version do you get this false positive? It doesn't reproduces for me using Python 2.7 and pylint 1.9.2.

**Comment by hvelarde** [comment]
yes, it's 1.9.2 and I can reproduce the issue locally:

```console
$ python -V
Python 2.7.14
$ grep pylint bin/pylint 
  '/home/hvelarde/.buildout/eggs/pylint-1.9.2-py2.7.egg',
from pylint import run_pylint
run_pylint()
$ bin/pylint --py3k --disable=no-absolute-import src/brasil/gov/agenda
No config file found, using default configuration
************* Module brasil.gov.agenda.utils
W: 89,48: xrange built-in referenced (xrange-builtin)
************* Module brasil.gov.agenda.content.agendadiaria
W:136,41: basestring built-in referenced (basestring-builtin)
************* Module brasil.gov.agenda.browser.ics
W:  6, 0: Module moved in Python 3 (bad-python3-import)
************* Module brasil.gov.agenda.tests.test_setup
W:189,29: Using a variable that was bound inside a comprehension (comprehension-escape)
W:196,29: Using a variable that was bound inside a comprehension (comprehension-escape)
W:203,29: Using a variable that was bound inside a comprehension (comprehension-escape)
W:212,29: Using a variable that was bound inside a comprehension (comprehension-escape)
W:219,29: Using a variable that was bound inside a comprehension (comprehension-escape)
************* Module brasil.gov.agenda.tests.test_agenda_tile
W: 83,12: Called a next() method on an object (next-method-called)

------------------------------------------------------------------
Your code has been rated at 9.96/10 (previous run: 9.97/10, -0.01)

```

**Comment by hvelarde** [comment]
@PCManticore in fact the error is in the line _after_ the list comprehension as you can see in the commits referenced.

**Comment by PCManticore** [comment]
@hvelarde I see where the issue is, the fact that you're using the same variable name for the result of the comprehension and for the iteration step, which is tripping pylint.

**Comment by jklemm** [comment]
So, I'm using pylint 2.2.2 and still getting this error.  😢

**Comment by Pierre-Sassoulas** [comment]
Won't do, we removed the python3 porting mode in 2.11.
